src/alpaka/AlpakaCore/alpakaWorkDivHelper.h

```diff
@@ -29,50 +29,268 @@ namespace cms {
+        // Take into account the block index in grid.
+        const uint32_t blockIdxInGrid(alpaka::idx::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[dimIndex]);
+        const uint32_t blockDimension(alpaka::workdiv::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[dimIndex]);
+
+        // Shift to get global indices in grid (instead of local to the block)
+        elementIdxShift[dimIndex] += blockIdxInGrid * blockDimension;
+      }
+
+      // Return element indexes, shifted by elementIdxShift.
+      return element_index_range_in_block(acc, elementIdxShift);
+    }
+
+    /*
+     * Computes the range of the elements indexes in grid.
+     * Truncated by the max number of elements of interest.
+     */
+    template <typename T_Acc, typename T_Dim>
+    ALPAKA_FN_ACC std::pair<Vec<T_Dim>, Vec<T_Dim>> element_index_range_in_grid_truncated(
+        const T_Acc& acc, const Vec<T_Dim>& maxNumberOfElements, Vec<T_Dim>& elementIdxShift) {
+      // Check dimension
+      static_assert(alpaka::dim::Dim<T_Acc>::value == T_Dim::value,
+                    "Accelerator and maxNumberOfElements need to have same dimension.");
+      auto&& [firstElementIdxGlobalVec, endElementIdxGlobalVec] = element_index_range_in_grid(acc, elementIdxShift);
+
+      // Truncate
       for (typename T_Dim::value_type dimIndex(0); dimIndex < T_Dim::value; ++dimIndex) {
         endElementIdxGlobalVec[dimIndex] = std::min(endElementIdxGlobalVec[dimIndex], maxNumberOfElements[dimIndex]);
       }
 
+      // Return element indexes, shifted by elementIdxShift, and truncated by maxNumberOfElements.
       return {firstElementIdxGlobalVec, endElementIdxGlobalVec};
     }
 
+    /*
+     * Computes the range of the element(s) index(es) in grid.
+     * Truncated by the max number of elements of interest.
+     */
+    template <typename T_Acc, typename T_Dim>
+    ALPAKA_FN_ACC std::pair<Vec<T_Dim>, Vec<T_Dim>> element_index_range_in_grid_truncated(
+        const T_Acc& acc, const Vec<T_Dim>& maxNumberOfElements) {
+      Vec<T_Dim> elementIdxShift = Vec<T_Dim>::zeros();
+      return element_index_range_in_grid_truncated(acc, maxNumberOfElements, elementIdxShift);
+    }
+
+    /*********************************************
+     *     1D HELPERS, LOOP ON ALL CPU ELEMENTS
+     ********************************************/
+
+    /*
+     * Loop on all (CPU) elements.
+     * Elements loop makes sense in CPU case only. In GPU case, elementIdx = firstElementIdx = threadIdx + shift.
+     * Indexes are local to the BLOCK.
+     */
+    template <typename T_Acc, typename Func>
+    ALPAKA_FN_ACC void for_each_element_in_thread_1D_index_in_block(const T_Acc& acc,
+                                                                    const uint32_t maxNumberOfElements,
+                                                                    const uint32_t elementIdxShift,
+                                                                    const Func func) {
```

Comment:
I think this should be either `const& Func` (as it was) or `Func&&`. It being `const` prevents mutating functors (which I'd think to be fine), and taking it by value implies a copy (which is not needed).

Reply:
Hmm passing by value would only do a copy if the argument was an lvalue?
Here in practice, the func arguments are actually always temporaries instead (though, they could also have been lvalues).
Don't know, in practice passing the lambda by value clearly leads to better performance than passing by `const &`, or passing by `&&` and then forwarding.

Reply:
Passing the lambda by && and then forwarding, and inlining these functions, seems to be a good compromise, will go for that.

Reply:
By quick look (that I should have done earlier), `std` and SYCL seem to take such functors by "non-const value", and Kokkos and Alpaka by "const reference". Just do double check, did you observe that passing the lambda by value gives better performance than reference?

If I've understood correctly, calling `std::forward()` many times on the same object is not strictly speaking safe (has the same issue as `std::move()`), even if it may work in practice in many cases.

Sorry for going back-and-forth, how about going back to the by-value (but without `const`), or keep the universal reference and remove the `std::forward`?

Reply:
Re-looking better at this, I do not see clear diff in perf.

- By value still makes the assumption that the function is light, which should be the case most of the time, but is not that nice.

- Func&& without forwarding loses a bit of interest, as the parameter (even a temporary) would be bound to an lvalue, no?

Bah maybe this is too much hair-pulling
> I think this should be either const Func& (as it was) or Func&&
I think I could revert to the initial implementation with `const&` ?

Also, as this does not change the helper functions interface anyway, maybe we could accept this, move on, and re-evaluate later when the full track reconstruction has been ported to have a better metric on the impact on perf?

Reply:
> * By value still makes the assumption that the function is light, which should be the case most of the time, but is not that nice.

I agree, although that is the precedent that `std` sets (not that we should always agree with `std`).

> * Func&& without forwarding loses a bit of interest, as the parameter (even a temporary) would be bound to an lvalue, no?

Within the function body the `func` of `Func&& func` would indeed be an lvalue. But the function argument itself would bind to anything without requiring the the `func` to be const.

> Bah maybe this is too much hair-pulling
> 
> > I think this should be either const Func& (as it was) or Func&&
> 
> I think I could revert to the initial implementation with `const&` ?
>
> Also, as this does not change the helper functions interface anyway, maybe we could accept this, move on, and re-evaluate later when the full track reconstruction has been ported to have a better metric on the potential impact on perf?

I agree, we should move on. I'm going to run tests for this and merge then. (although the initial implementation was/is `const Func` instead of `const Func&`)